### ISSUE TITLE
Nesting Kokkos::View initialization

### ISSUE THREAD

**Issue by keitat** [kokkos/kokkos#1254]
I am having runtime error with initialization of Kokkos::View (OpenMP backend on Intel platform).
In my program, I have a statement like:
`
Kokkos::View<MyClass>  myView("View",10);
`
MyClass is a relatively complex class that contains many class members, some of which are Kokkos::View. 

In constructor,  these internal Kokkos::View are initialized like:
`
MyClass() : myView1("anotherView",1), myView2( ...) , myView( ..) 
{
 // Initializing ....
}
`
When I tested the code with 1 thread, the code works fine.  However, I am having the following errors:

>terminate called after throwing an instance of 'std::runtime_error'
> what():  Constructing View and initializing data with uninitialized execution space
> Traceback functionality not available

or 

>terminate called after throwing an instance of 'terminate called recursively
> std::runtime_error'

I thought that my Kokkos usage is legal (no compile time error).  Is it a bug in Kokkos?

**Comment by crtrott** [comment]
The default constructor of a class you use as value_type for Kokkos::View is not allowed to allocate data, which includes calling an allocating constructor of a Kokkos::View. Basically allocating data is not supported by some parallel execution spaces hence, we don't allow you to do that for any to have portability. What you need to do is for your default constructor to only call the default constructors of the views, and then have some kind of init function which actually allocates the views. For full portability you need to do something like this:

```c++
View<MyClass*> my_stuff("MyStuff",10);
View<MyClass*>::HostMirror h_my_stuff = create_mirror_view(my_stuff);
for(int i=0; i<10; i++) 
  h_my_stuff.init();
deep_copy(my_stuff,h_my_stuff);
```

Unfortunately you need to keep h_my_stuff alive until my_stuff goes out of scope, for complicated reasons. What I tend to do is put the View<MyClass*> view into a memory space which allows access from the host. Then you don't need the host copy of stuff, and all the reference counting works more like what you expect.

**Comment by mhoemmen** [comment]
This needs to go into an FAQ so we can just point people to it -- @crtrott would you mind if I were to start a wiki entry?

**Comment by crtrott** [comment]
Mark feel free to do so. We need to modify 6.2.2 and then probably add a 6.6 Views of Classes.
Feel free to sync the wiki back to the wiki in the ProgrammingGuide repo. If you write significant addition we probably need to get those additions through R&A before we can merge it into the public wiki.

**Comment by crtrott** [comment]
Ok I merged the public wiki back into the private one. You should have write access to the latter one so feel free to add stuff there.

**Comment by mhoemmen** [comment]
@crtrott I just added a pull request to the private wiki: https://github.com/kokkos/ProgrammingGuide/pull/21 (should only be accessible to members).
